lib/connect.ts

```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
+    login_uri?: string;
+    debug?: boolean;
+    timeout?: number;
+    default_connect_uri?: string;
+    session_expiry_hook?: string | null;
+    default_login_uri?: string;
+    private requestInstance: AxiosInstance;
+
+    constructor(params: KiteConnectParams) {
+        this.api_key = params.api_key;
+        this.root = params.root || DEFAULTS.root;
+        this.timeout = params.timeout || DEFAULTS.timeout;
+        this.debug = params.debug || DEFAULTS.debug;
+        this.access_token = params.access_token || null;
+        this.default_connect_uri = DEFAULTS.login;
+        this.session_expiry_hook = null;
+        this.requestInstance = this.createAxiosInstance();
+    }
+
+    private createAxiosInstance(): AxiosInstance {
+        const kiteVersion = 3; // Kite version to send in header
+        const userAgent = utils.getUserAgent();  // User agent to be sent with every request
+        const requestInstance = axios.create({
+            baseURL: this.root as string,
+            timeout: this.timeout,
+            headers: {
+                'X-Kite-Version': kiteVersion,
+                'User-Agent': userAgent
+            },
+            paramsSerializer(params) {
+                return querystring.stringify(params);
+            }
+        });
+
+        // Add a request interceptor
+        requestInstance.interceptors.request.use((request) => {
+            if (this.debug) console.log(request);
+            return request;
+        });
+
+        // Add a response interceptor
+        requestInstance.interceptors.response.use((response) => {
+            if (this.debug) console.log(response);
+
+            const contentType = response.headers['content-type'];
+            if (contentType === 'application/json' && typeof response.data === 'object') {
+                // Throw incase of error
+                if (response.data.error_type) throw response.data;
+
+                // Return success data
+                return response.data.data;
+            } else if (contentType === 'text/csv') {
+                // Return the response directly
+                return response.data
             } else {
-                resp.error_type = "NetworkException";
-                resp.message = error.response.statusText;
+                return {
+                    'error_type': 'DataException',
+                    'message': 'Unknown content type (' + contentType + ') with response: (' + response.data + ')'
+                };
             }
-        } else if (error.request) {
-            // The request was made but no response was received
-            // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
-            // http.ClientRequest in node.js
-            resp.error_type = "NetworkException";
-            resp.message = "No response from server with error code: " + error.code;
-        } else if (error.message) {
-            resp = error
-        }
+        }, function (error) {
```

Comment:
Missed specifying the type of error, either you can define `(error: Error)` as the desired type. If you want to allow any type for the error parameter, you can use `(error: any)`